Timings are now done with a AsyncTask.

madecode15 opened the pull request:
This PR allows timings to paste on a separate thread.

Why? It took 302~ ms to do it on the main thread for me, and that causes big lag spikes on a production server, where you would probably want to get timings from.

Issues:
- Not sure how to propery print exceptions on a different thread, it would be bad to access the logger on it... and you can't get the exception back to the main thread.
- Translation needed for "Processing timings...".

dktapps, let me know how you would like stuff refactored.

Otherwise all works well.

Commit:
Timings are now done with a AsyncTask.

dktapps commented on src/pocketmine/command/defaults/Threads/TimingsPasteTask.php:
```diff
@@ -0,0 +1,80 @@
+
+	public function __construct($name, $data, $serverName, $version){
+		$this->name = $name;
+		$this->data = unserialize($data);
```
why serializing/unserializing here? you're still on the main thread in the constructor. You should be _serializing_ here, not unserializing. An AsyncTask is not a thread.

dktapps commented on src/pocketmine/command/defaults/Threads/TimingsPasteTask.php:
```diff
@@ -0,0 +1,80 @@
+		curl_setopt($ch, CURLOPT_SSL_VERIFYHOST, 2);
+		curl_setopt($ch, CURLOPT_FORBID_REUSE, 1);
+		curl_setopt($ch, CURLOPT_FRESH_CONNECT, 1);
+		curl_setopt($ch, CURLOPT_POSTFIELDS, $this->data);
```
Likely to run into issues here due to the serialization issue above. As a result this will be a Threaded object, not an array.

dktapps commented:
I had already intended to do this myself, you're right, it does take a signficant amount of time to execute. I don't know why it wasn't asynchronous to begin with.

Commit:
Remove not needed serialize/unserialize.

madecode15 commented:
Shoghicp is not a fan big fan of threading for some reason. Well I don't blame him, phthreads is a pain in the ass to work with and the OOP style coding does not help with that either.

madecode15 commented:
Whoops.. Found a issue where it does not let you choose what host you want. Will fix that soon.

Commit:
Added back timings host selection option... Also correctly use "timings.pmmp.io" as the default instead of mcpetimings.

SOF3 commented on src/pocketmine/Server.php:
```diff
@@ -1801,6 +1801,17 @@ public function broadcastPacketsCallback($data, array $identifiers){
+			$sender->sendMessage(new TranslationContainer("pocketmine.command.timings.timingsUpload", ["http://paste.ubuntu.com/" . $matches . "/"]));
+			$sender->sendMessage(new TranslationContainer("pocketmine.command.timings.timingsRead", ["http://" . $this->getProperty("timings.host", "timings.pmmp.io") . "/?url=" . $matches]));
+		}
+	}
```
I don't like having this callback method here. Putting everything into the Server class is not favourable in the long term.

madecode15 replied:
Yeah I was not a fan of it either... That is why I was wondering where dktapps/you would like it.

robske110 replied:
maybe actually in the command? Because a timing paste is only started from there. The current timings classes are not suitable for this method for sure.

SOF3 replied:
Maybe use `fetchLocal` with the command instance to callback?

SOF3 commented on src/pocketmine/Server.php:
```diff
@@ -1801,6 +1801,17 @@ public function broadcastPacketsCallback($data, array $identifiers){
 	}
 
+	public function timingsPasteCallback($name, $matches, $failed){
+		$sender = $name === "CONSOLE" ? $this->consoleSender : $this->getPlayerExact($name);
```
Ever considered using `AsyncTask::fetchLocal()`?

SOF3 commented on src/pocketmine/command/defaults/TimingsCommand.php:
```diff
@@ -102,37 +103,10 @@ public function execute(CommandSender $sender, $currentAlias, array $args){
+
+				$sender->getServer()->getScheduler()->scheduleAsyncTask(new TimingsPasteTask($sender->getName(), $data, $sender->getServer()->getName(), $sender->getServer()->getPocketMineVersion()));
+
 				fclose($fileTimings);
```
Just for code clarification that you aren't trying to use $fileTimings in TimingsPasteTask, it would look more beautiful if you place this line above the schedule line. This is not necessary, just suggesting that this be done (actually should have been done since @shoghicp did it)

ICrashHD commented:
www.pornhub.com/pornstar/leah-gottiwww.pornhub.com/pornstar/leah-gottiwww.pornhub.com/pornstar/leah-gottiwww.pornhub.com/pornstar/leah-gottiwww.pornhub.com/pornstar/leah-gottiwww.pornhub.com/pornstar/leah-gottiwww.pornhub.com/pornstar/leah-gottiwww.pornhub.com/pornstar/leah-gottiwww.pornhub.com/pornstar/leah-gottiwww.pornhub.com/pornstar/leah-gottiwww.pornhub.com/pornstar/leah-gottiwww.pornhub.com/pornstar/leah-gottiwww.pornhub.com/pornstar/leah-gottiwww.pornhub.com/pornstar/leah-gottiwww.pornhub.com/pornstar/leah-gotti

Commit:
Attempt to rectify SOF3's requested changes.

madecode15 commented:
@SOF3 

Requesting new review to apprehend with latest commit.

Changes:
- Moved fclose() to match up with PocketMine's style.
- Removed timingspastecallback from the server class.
- Added a getter for consoleCommandSender, because there was no way to access it.
- Optimized code (not important)

SOF3 commented on src/pocketmine/Server.php:
```diff
@@ -2291,6 +2290,13 @@ public function getMemoryManager(){
+	 * @return ConsoleCommandSender
+	 */
+	public function getConsoleCommandSender(){
+		return $this->consoleSender;
```
again, consider using fetchLocal.

madecode15 replied:
Done.

Commit:
Use fetch local and other great API methods

madecode15 commented:
Closing for now since making a generic task for curl operations is a better solution.
